File: frontend/src/types/custom-agent.ts
export type CustomAgentStatus = "active" | "promoted" | "archived";

export interface CustomAgent {
  id: string;
  name: string;
  description: string;
  category: string;
  tags: string[] | null;
  source_query: string;
  status: CustomAgentStatus;
  try_count: number;
  completion_count: number;
  avg_rating: number;
  upvote_count: number;
  promoted_agent_id: string | null;
  created_by_user_id: string | null;
  created_at: string;
  updated_at: string;
  user_vote: number | null;
}

export interface CustomAgentListResponse {
  agents: CustomAgent[];
  total: number;
}

export interface CreateAgentRequest {
  message: string;
  category?: string;
  auto_execute?: boolean;
}

export interface CreateAgentResponse {
  agent: CustomAgent;
  task_id: string | null;
  task_status: string | null;
  result: string | null;
}

export interface TryAgentRequest {
  message: string;
}

export interface VoteRequest {
  vote: number;
}

export interface AgentRequest {
  id: string;
  user_id: string | null;
  query: string;
  best_match_confidence: number | null;
  custom_agent_id: string | null;
  created_at: string;
}

export interface AgentRequestListResponse {
  requests: AgentRequest[];
  total: number;
}

File: src/schemas/custom_agent.py
"""Schemas for custom (community-created) agents."""

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field


class CreateAgentRequest(BaseModel):
    message: str = Field(max_length=10_000, description="What kind of agent do you need?")
    category: Optional[str] = Field(None, max_length=100)
    auto_execute: bool = Field(default=True, description="Run the first task immediately")


class TryAgentRequest(BaseModel):
    message: str = Field(max_length=10_000, description="Task message for the agent")


class VoteRequest(BaseModel):
    vote: int = Field(description="+1 or -1")


class CustomAgentResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str
    description: str
    category: str
    tags: list[str] | None = None
    source_query: str
    status: str
    try_count: int
    completion_count: int
    avg_rating: float
    upvote_count: int
    promoted_agent_id: str | None = None
    created_by_user_id: str | None = None
    created_at: datetime
    updated_at: datetime
    # Populated per-request for authenticated users
    user_vote: int | None = None


class CustomAgentListResponse(BaseModel):
    agents: list[CustomAgentResponse]
    total: int


class CreateAgentResponse(BaseModel):
    agent: CustomAgentResponse
    task_id: str | None = None
    task_status: str | None = None
    result: str | None = None


class AgentRequestResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    user_id: str | None
    query: str
    best_match_confidence: float | None
    custom_agent_id: str | None
    created_at: datetime


class AgentRequestListResponse(BaseModel):
    requests: list[AgentRequestResponse]
    total: int
